src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -336,44 +946,81 @@ func (s *commitLogSource) shouldIncludeInIndex(
 	return rangesToBootstrap.Overlaps(indexBlockRange)
 }
 
-func (s *commitLogSource) mergeShards(
+func (s *commitLogSource) mergeAllShardsCommitLogEncodersAndSnapshots(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	snapshotFiles map[uint32]fs.FileSetFilesSlice,
+	mostRecentCompleteSnapshotByBlockShard map[xtime.UnixNano]map[uint32]fs.FileSetFile,
 	numShards int,
-	bopts result.Options,
 	blockSize time.Duration,
-	blopts block.Options,
-	encoderPool encoding.EncoderPool,
 	unmerged []shardData,
-) result.DataBootstrapResult {
+) (result.DataBootstrapResult, error) {
 	var (
-		shardErrs               = make([]int, numShards)
-		shardEmptyErrs          = make([]int, numShards)
-		bootstrapResult         = result.NewDataBootstrapResult()
-		blocksPool              = bopts.DatabaseBlockOptions().DatabaseBlockPool()
-		multiReaderIteratorPool = blopts.MultiReaderIteratorPool()
+		shardErrs       = make([]int, numShards)
+		shardEmptyErrs  = make([]int, numShards)
+		bootstrapResult = result.NewDataBootstrapResult()
 		// Controls how many shards can be merged in parallel
 		workerPool          = xsync.NewWorkerPool(s.opts.MergeShardsConcurrency())
 		bootstrapResultLock sync.Mutex
 		wg                  sync.WaitGroup
 	)
-
 	workerPool.Init()
 
 	for shard, unmergedShard := range unmerged {
+		snapshotData, err := s.bootstrapShardSnapshots(
+			ns.ID(),
+			uint32(shard),
+			false,
+			shardsTimeRanges[uint32(shard)],
+			blockSize,
+			snapshotFiles[uint32(shard)],
+			mostRecentCompleteSnapshotByBlockShard,
+		)
+		if err != nil {
+			bootstrapResultLock.Lock()
+			// Mark the shard time ranges as unfulfilled so a subsequent bootstrapper
+			// has the chance to fulfill it.
+			bootstrapResult.Add(
+				uint32(shard),
+				result.NewShardResult(0, s.opts.ResultOptions()),
+				shardsTimeRanges[uint32(shard)],
+			)
+			bootstrapResultLock.Unlock()
+			return nil, err
+		}
+
 		if unmergedShard.series == nil {
-			continue
+			if !snapshotData.IsEmpty() {
+				// No snapshot or commit log data, skip.
+				// TODO: When we start allowing peer bootstrapping after commit log bootstrapping,
+				// then we will need to change this to mark the shard time ranges as unfulfilled.
```

Comment:
Hm any reason not to do this now?

Reply:
I think this was an unclear comment, its more like just in general we'll need to be smarter about  whether or not the existence or absence of snapshot files constitutes success (with no data) or unfulfilled. I think we discussed this briefly and we were talking about using the shard initialization state. That change alone will probably warrant a one page design doc